### ISSUE TITLE
Accessibility for Users who are Blind

### ISSUE THREAD

**Issue by johapatro** [supercollider/supercollider#790]
Please, will someone make Supercollider accessible for users who are blind? I recently taught a class on music and songwriting, and my students thoroughly enjoyed a demonstration of Supercollider for creating electronic music. However, I tested it today, and it does not seem to function with Voiceover. This could truly broaden the horizons of a number of users who are visually impaired, so if there is anyone who could make this happen, I would love to start a dialogue.
Thank you,
John Rose
Short-Term Programs Teacher
Texas School for the Blind and Visually Impaired

**Comment by vivid-synth** [comment]
@johapatro this is an old issue but if you'd like to comment, you weren't tagged earlier.

**Comment by amtlab** [comment]
Hello. I tested 3.8 on Mac with Sierra. It seems like VO can read some elements on the GUI.
However the main problem I ran into right away is that SC just freezes when I go to text input area and starts typing. Here are the steps you can follow to reproduce.
1. Press command+f5 to turn on/off voice over
2. Go to SC with Voiceover on
3. Move VO cursor to the input area by pressing control+option+right/left arrows.
4. Once you start typing, SC should freeze and wheel until you force quit.
As long as you avoid the input area, VO seems to be able to read the console output, and so on.
Any thought?
Thanks!

**Comment by brianlheim** [comment]
More user info: I just tested current dev build of 3.9 on macOS 10.11 and VO works fine for me.

**Comment by amtlab** [comment]
Hmm, I just tried the latest build from
http://supercollider.s3.amazonaws.com/builds/supercollider/supercollider/osx/master-latest.html
Unfortunately it still freezes when I start typing and edit.
Could someone on Sierra confirm?
Thanks!

**Comment by brianlheim** [comment]
Now that I've updated to Sierra, I tried testing again. VoiceOver still works fine for me. However, I can get it to freeze in the Help Browser, which seems to be unnavigable.

**Comment by jsl303** [comment]
Same. It just freezes when I type something in the area where Voiceover says edit text. Just need to force quit if that happens. Hope it gets fixed soon. I'd love to be able to use it!

**Comment by brianlheim** [comment]
@jsl303 Sorry, just to clarify, is the place where it freezes the main text editor, or the editable text search field in the help browser?

**Comment by jsl303** [comment]
Brian, actually I'm not sure.

I installed, opened, Tried to navigate by pressing vo+right, then and I got

Untitle SuperCollider IDE

Untitled group

Menu button

Help browser text

Text

Home button

Go back dimmed button

Go forward dimmed button

Reload button

Edit text blank

Edit text

Then it freezes.

Is that help area? If so, is there way to hide it?

Thanks!

**Comment by brianlheim** [comment]
Yes, that is the help area. Here's how you can close it for now:

vo+m (focus menu bar)

right to "view"

down to "docklets"

right, then down to "help browser"

press enter to close the help docklet.

It seems like this might be an issue with how elements are being grouped in the main GUI window...

**Comment by jsl303** [comment]
Brian, that worked! No more freeze in the main window!!!

FYI, same thing happens in the preference in general and shortcuts tabs.
Hope it can get fixed in the future.

**Comment by brianlheim** [comment]
I'm glad it helped! Thanks for the info about the preferences freezes.

3.9 is going to unofficially be the "internationalization and cross-platform-ization" release, and I'm going to push to get this solved for either 3.9.0 or 3.9.1.

**Comment by widp** [comment]
There's also http://emacspeak.sourceforge.net/ for emacs. I am not sure about its interface with supercollider.
.

**Comment by syntheticuniverse** [comment]
Hi!

I found an accessibility bug in SuperCollider 10.3.2. System is Mac Os 10.14.5. The help browser doesn't hang anymore, but Voiceover does not read text in the editor, when navigating with up and down arrows. Left and right works, so i can navigate by character or by word with CTRL+Opt+left and right.

Can anyone confirm this or is there a workaround?

Greetings from a newbie :-)

**Comment by syntheticuniverse** [comment]
Sorry, I mean SuperCollider 3.10.2

**Comment by RHKremser** [comment]
Hi everybody!
I am relatively new to SuperCollider, and have similar experiences.
my system: SuperCollider (latest MacOs release, Catalina latest, Mac Mini i7 gen8.
Navigating the edit window works as described by . VO+arrow left or right is as usual, VO+arrow up/down sometimes works as expected, sometimes it does not: I feel, the longer I use these commands, the more often they do not work any more and respond via "Beep" (like not possible), if using just the arrow keys, this works more often, but it leads to some confusion with the VO focus, especially in long code files.
I could not figure out, how to navigate to explicit lines in code, though the option is marked in the menu.

But this is not bad anyway, writing code, evaluating it and getting results do not collaborate perfectly with VoiceOver, but Apple stock software does not work this way to all extends either.
SuperCollider help system seems to me a more tricky problem.
Once in the help panel, some options do not work when trying to set actions with common VO shortcuts:
Links do not open by VO +space, BUT by mouse replacement double click.
Some actions are not reported by VoiceOver (also do not show up in the items list):
I tried to get through with the assistance of a seeing person, but I got lost very soon.
I only see a solution like this: a seeing person who is familiar with SuperCollider help system is asked to check the functionality by using VoiceOver, and to provide some tutorial especially for the needs of blind and visually disabled. I myself can imagine the efficiency of the help system only if i.e. Eli Fieldsteel explains it in his precise manner, but using VoiceOver everything is different, especially if someone is used i.e. to Apple's help system or similar ones and not to a help system that is related to programming environments.
I hope, someone kindly does this effort. If yes, many thanks this way!
best regards
Rainer
